src/libavrdude.i

```diff
@@ -0,0 +1,561 @@
+void set_progress_callback(PyObject *PyFunc);
+
+// These things are read from config file(s), and must be considered
+// read-only by any program. Most internals are only relevant for
+// libavrdude itself, so only map things to the Python level that are
+// needed there.
+
+%typemap(out) unsigned char signature[3] {
+  $result = PyBytes_FromStringAndSize((const char *)$1, 3);
+ }
+
+%immutable;
+typedef struct avrpart {
+  const char  * desc;               /* long part name */
+  const char  * id;                 /* short part name */
+  LISTID        variants;           /* String with variant name and chip properties */
+  const char  * parent_id;          /* Used by developer options */
+  const char  * family_id;          /* family id in the SIB (avr8x) */
+  int           prog_modes;         /* Programming interfaces, see #define PM_... */
+  unsigned char signature[3];       /* expected value of signature bytes */
+  unsigned short usbpid;            /* USB DFU product ID (0 = none) */
+  LISTID        mem;            /* avr memory definitions */
+  LISTID        mem_alias;      /* memory alias definitions */
+  const char  * config_file;    /* config file where defined */
+  int           lineno;         /* config file line number */
+} AVRPART;
+%mutable;
+
+typedef unsigned int memtype_t;
+typedef struct avrmem {
+  %immutable;
+  const char *desc;           /* memory description ("flash", "eeprom", etc) */
+  memtype_t type;             /* internally used type, cannot be set in conf files */
+  bool paged;                  /* 16-bit page addressed, e.g., ATmega flash but not EEPROM */
+  int size;                   /* total memory size in bytes */
+  int page_size;              /* size of memory page (if page addressed) */
+  int num_pages;              /* number of pages (if page addressed) */
+  int initval;                /* factory setting of fuses and lock bits */
+  int bitmask;                /* bits used in fuses and lock bits */
+  %mutable;
+  unsigned char * buf;        /* pointer to memory buffer */
+} AVRMEM;
+
+%extend avrmem {
+  PyObject *get(unsigned int len, unsigned int offset = 0) {
+    if (offset >= (unsigned)$self->size)
+      return Py_None;
+    if (offset + len > (unsigned)$self->size)
+      len = $self->size - offset;
+    return PyBytes_FromStringAndSize((char *)($self->buf + offset), len);
+  }
+}
+
+int avr_initmem(const AVRPART *p);
+
+%extend avrmem {
+  %typemap(in) (unsigned char *in, unsigned int len) {
+    Py_ssize_t len;
+    PyBytes_AsStringAndSize($input, (char **)&$1, &len);
+    $2 = len;
+  }
+  int put(unsigned char *in, unsigned int len, unsigned int offset = 0) {
+    if ($self->buf == NULL)
+      // missing avr_initmem()?
+      return 0;
+    if (offset >= (unsigned)$self->size)
+      return 0;
+    if (offset + len > (unsigned)$self->size)
+      len = $self->size - offset;
+    memcpy($self->buf + offset, in, len);
+    memset($self->tags + offset, TAG_ALLOCATED, len);
```

Comment:
And how can tags be cleared again? Should there be a clear method?

Reply:
> And how can tags be cleared again? Should there be a clear method?

Do you really think it's necessary? ;-)

I guess reading a memory ought to clear the tag … anything else?

Reply:
Actually, I was assuming the read operation would clear any tag.

Reply:
I tried to find the places where regular avrdude / libavrdude code actually clears the tags. I seen only two places:

* during initialization (obviously)
* in `fileio_segments_normalise`

Of course, the Python wrapper should expose the public functions from `fileio.c`. That would be one option then (the same as for the CLI) to clear the tags.
However, shouldn't also any memory read operation clear the tag? The data come from the device then, so they are not supposed to be tagged as "ALLOCATED" anymore.
Bottom line: I'm also going to supply a method to "clear" certainly `AVRMEM*` regions from Python.

Reply:
The `tags[]` array is set when reading input from file, to inform the
 - Write-to-device routine(s) which data are meant to transfer to the device 
 - Verify routine(s) which data are meant to be read from the device for verification

That's all there is. `tags[]` is cleared before file read: happens, as you observed, in `fileio_segments_normalise()`

In C, the user of `libavrdude` can read a file and then manipulate the contents to their heart (eg, to implement an interactive editor) b/c the `tags[]` array is exposed in its full glory. Without the Python clear method, a `libavrdude` user couldn't easily implement such an interactive editor.

> shouldn't also any memory read operation clear the tag?

No: that was never the case and, if implemented, might surprise some use cases that rely on current behaviour.